### PR TITLE
Cat unrandomize game

### PR DESCRIPTION
**Pull request by Cattharine**
The randomness of the game can interfere with accurate evaluation of models. So it was decided to somewhat "unrandomize" game. There is two parts of randomness in our evaluations: the game itself and our card picking for creating state of environment.
This PR adds ability to get rid of both

### CODE REVIEW

File: environment/reward_sytem/potential_end_stage_reward_system.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from .base_reward_system import BaseRewardSystem
+
+
+class PotentialEndStageRewardSystem(BaseRewardSystem):
```
Comment (krutovsky-danya) [inline environment/reward_sytem/potential_end_stage_reward_system.py:4]: Ух, пролезли изменения из другого PR 😑

File: full_main.py [hunk 2]
```diff
@@ -31,7 +31,7 @@
     # eval model
     # collect quality metrics
 
-# analyze collected metrics
+# analize collected metrics
```
Comment (krutovsky-danya) [inline full_main.py:34]: Heh, my bad. Можно тоже в отдельный PR выносить такое)

File: game/common_methods.py [hunk 3]
```diff
@@ -35,15 +36,17 @@ def interpolate(value, table: dict):
...
     return None
 
+
 def sample_n_or_less(collection, count):
     count = min(count, len(collection))
     return random.sample(collection, count)
 
-def sample_n_or_zero(collection: Sequence, count: int):
+
+def sample_n_or_zero(collection: Sequence, count: int, gen: np.random.Generator):
```
Comment (krutovsky-danya) [inline game/common_methods.py]: Почему имя именно gen, а не generator?

File: game/game.py [hunk 4]
```diff
@@ -13,13 +13,12 @@
...
 from game.common_methods import interpolate, stepify, clamp
 from game.rooms.devroom.room import get_current_room_cost, get_worker_cost
-import random
 from typing import List, Dict
 
 
 class ProductOwnerGame:
-    def __init__(self):
-        self.context = GlobalContext()
+    def __init__(self, seed=None, card_picker_seed=None):
```
Comment (krutovsky-danya) [inline game/game.py]: Может быть сразу оперировать генераторами?

File: game/game.py [hunk 5]
```diff
@@ -181,44 +180,48 @@ def _update_profit_ordinary_card(self, us: UserStoryCardInfo):
...
-                ran_usr_to_bring = us.customers_to_bring * \
-                    random.uniform(us_fp[0], us_fp[1])
+            is_last_key = us_fp_key == GlobalConstants.sorted_keys_userstory_floating_profit[-1]
+            if sprints_spent <= us_fp_key or is_last_key:
+                ran_val = self.context.random_gen.uniform(us_fp[0], us_fp[1])
+                ran_usr_to_bring = us.customers_to_bring * ran_val
                 self.context.customers += stepify(ran_usr_to_bring, 0.01)
-                ran_lty_to_bring = us.loyalty * \
-                    random.uniform(us_fp[0], us_fp[1])
+                ran_val = self.context.random_gen.uniform(us_fp[0], us_fp[1])
```
Comment (krutovsky-danya) [inline game/game.py]: Раз мы уже здесь, давай красиво назовем эти переменные?
Очень уж сильно пришлось подумать, что такое ran, хоть и PR про random

File: game/game.py [hunk 6]
```diff
@@ -13,13 +13,12 @@
...
 from game.rooms.devroom.room import get_current_room_cost, get_worker_cost
-import random
 from typing import List, Dict
 
 
 class ProductOwnerGame:
-    def __init__(self):
-        self.context = GlobalContext()
+    def __init__(self, seed=None, card_picker_seed=None):
+        self.context = GlobalContext(seed=seed, card_picker_seed=card_picker_seed)
```
Comment (krutovsky-danya) [inline game/game.py]: А получится не добавлять новые поля в GlobalContext?
Общий контекст - анти паттерн, давай не будем его развивать)

File: game/game_variables.py [hunk 7]
```diff
@@ -32,6 +38,8 @@ def __init__(self) -> None:
         self.color_storage = ColorStorage()
         self.is_victory = False
         self.is_loss = False
+        self.random_gen = Random(x=seed) if seed is not None else Random()
```
Comment (krutovsky-danya) [inline game/game_variables.py]: Почему это разные генераторы случайных чисел?

File: game/userstories/userstories_generator.py [hunk 8]
```diff
@@ -11,12 +11,17 @@ def __init__(self, s: int, m: int, l: int, xl: int):
         for size in sizes_count:
             self.a += [size[1]] * size[0]
 
-    def generate_userstories(self, count: int, spawn_sprint: int, color_storage: ColorStorage):
+    def generate_userstories(self, count: int, spawn_sprint: int, color_storage: ColorStorage,
+                             random_gen: Random):
```
Comment (krutovsky-danya) [inline game/userstories/userstories_generator.py]: Может быть в конструкторе передавать генератор?
Мне нравится и так и так. В конструкторе выглядит даже логичнее, потому что это генератор, который использует другой генератор

File: game/userstories/userstories_generator.py [hunk 9]
```diff
@@ -11,12 +11,17 @@ def __init__(self, s: int, m: int, l: int, xl: int):
...
-    def generate_userstories(self, count: int, spawn_sprint: int, color_storage: ColorStorage):
+    def generate_userstories(self, count: int, spawn_sprint: int, color_storage: ColorStorage,
+                             random_gen: Random):
         result = []
 
         for i in range(count):
-            card_type = self.a[random.randint(0, len(self.a) - 1)]
-            card = UserStoryCard(UserStoryCardInfo(card_type, spawn_sprint, color_storage))
+            random_index = random_gen.randint(0, len(self.a) - 1)
+            card_type = self.a[random_index]
```
Comment (krutovsky-danya) [inline game/userstories/userstories_generator.py]: Это же choice!

File: pipeline/base_study.py [hunk 10]
```diff
@@ -52,9 +52,9 @@ def _choose_action(self, action, inner_sprint_action_count) -> Tuple[int, int]:
             inner_sprint_action_count = 0
         return action, inner_sprint_action_count
 
-    def study_agent(self, episode_n: int):
+    def study_agent(self, episode_n: int, seed=None, card_picker_seed=None):
         for episode in range(episode_n):
-            state = self.env.reset()
+            state = self.env.reset(seed=seed, card_picker_seed=card_picker_seed)
```
Comment (krutovsky-danya) [inline pipeline/base_study.py:57]: Почему seed не сохранить во внутренне поле? Тогда можно будет не менять reset)

File: environment/__init__.py [hunk 11]
```diff
@@ -4,7 +4,7 @@
 from .stochastic_game_start_env import StochasticGameStartEnv
 
 
-def make(is_vanilla=True):
+def make(is_vanilla=True, seed=None, card_picker_seed=None):
     if is_vanilla:
-        return ProductOwnerEnv()
-    return StochasticGameStartEnv()
+        return ProductOwnerEnv(seed=seed, card_picker_seed=card_picker_seed)
```
Comment (krutovsky-danya) [inline environment/__init__.py:9]: Я не понял, зачем нужно было фиксировать рандом в игре?

Если фиксировать рандом для оценки модели, то оценка будет не точной и не даст полного представления о качестве модели.

Если фиксировать для обучения, то будет еще хуже, агент будет видеть одну игру, а в вебе будет совсем другую.

А вот идея с построением распределения мне очень понравилась. Она как раз показывает больше информации о модели, которая получилась в итоге. Например, какой результат от нее ждать и в каком диапазоне)

Оставлю этот коммент здесь, чтобы можно было отвечать связно.

File: environment/reward_sytem/potential_end_stage_reward_system.py (hunk 1 above)
Comment (Cattharine) [reply environment/reward_sytem/potential_end_stage_reward_system.py:4]: Упс

File: full_main.py (hunk 2 above)
Comment (Cattharine) [reply full_main.py:34]: Их обычно не так много, чтобы заводить отдельный PR. Наверное

File: game/common_methods.py (hunk 3 above)
Comment (Cattharine) [reply game/common_methods.py]: Чтобы строчка не была слишком длинной. Если я поменяю тут gen на generator, придется менять везде, что приведет к увеличению длины строки в местах, где это будет выглядеть громоздко

File: game/game.py (hunk 4 above)
Comment (Cattharine) [reply game/game.py]: Зачем? Тогда их придется создавать там, откуда вызывается инициализация игры, а этого мне хотелось бы избежать

File: game/game.py (hunk 6 above)
Comment (Cattharine) [reply game/game.py:21]: Я не придумала ничего лучше. Из других вариантов либо прокидывать во все заинтересованные методы (тут ещё вопрос, где они, поля, тогда храниться будут), либо делать ещё более глобальными, что ещё хуже

File: game/game_variables.py (hunk 7 above)
Comment (Cattharine) [reply game/game_variables.py]: Я забыла, что хотела убрать этот if, оно и без него будет работать так же. Спасибо)

File: pipeline/base_study.py (hunk 10 above)
Comment (Cattharine) [reply pipeline/base_study.py:57]: seed полагается константой для отдельной игры, траектории, и не полагается константой для всей среды
Тогда и менять его придется через внутреннее поле. Тогда придется менять pipeline, везде перед reset вставить изменение seed. По-моему, это немного неудобно, есть шанс забыть о необходимости изменить seed

File: environment/__init__.py (hunk 11 above)
Comment (Cattharine) [reply environment/__init__.py:9]: 0. Это лишь возможность зафиксировать рандом. Всегда есть возможность выставить seed=None, и все будет работать, как раньше)

1. Цель у этого была такая: у меня есть базовая модель: сеточка для кредита + пустые спринты. Я хочу проверить, насколько хорошо новые модели: та же сеточка для кредита + агент для финальной части игры, играет относительно этой базовой модели. Но тут вмешивается рандом, где даже базовая модель имеет разброс где-то в 30% побед (при тесте на 100 траекториях). При этом новых моделей у меня около 100 с одного обучения. И у каждой при тесте на 100 траекториях тоже потенциально будет разброс в 30% побед. Больше траекторий - больше точность, но и больше траекторий - больше потраченного на оценку времени. Поэтому я захотела иметь возможность зафиксировать набор траекторий, на которых мне известно поведение базовой модели, и сравнить это поведение с поведением новых моделей. Из предположения, что если в новых моделях наобучалось что-то качественно отличное, то это будет заметно

2. Есть возможность зафиксировать несколько состояний выплаты кредита (1 состояние - 1 seed) и сравнить работу финальных моделей из одинаковых состояний. Это может быть более показательно, учитывая разброс возможных побед у одной модели кредита. Впрочем, я не отрицаю, что может попасться набор seed'ов, на которых какая-то потенциально хорошая модель будет показывать себя совсем плохо. Для моделей, стартующих с начала игры, фиксировать рандом смысла не имеет (только для воспроизводимости результата, разве что)

3. Для обучения не предполагается ничего фиксировать)

4. С этим согласна. Для какой-то конкретной модели построение такого распределения может быть очень даже полезно) Но использовать этот метод для сравнения большого количества моделей нецелесообразно, так как построение распределения - длительный процесс, если мы хотим получить достаточно точные данные

File: game/game.py (hunk 6 above)
Comment (Cattharine) [reply game/game.py]: Попыталась убрать новые поля из GlobalContext)

File: environment/__init__.py (hunk 11 above)
Comment (krutovsky-danya) [reply environment/__init__.py:9]: 0. Согласен, что это расширение функциональности, поэтому добить PR можно)

1. Кажется для этого хорошо подошло бы копирование среды. Это полезная функциональность, которая нужна в классических алгоритмах, типа Monte Carlo Tree Search

2. А для этого максимально подходит копирование среды, оно еще и быстрее будет) (Не надо нейронку гонять лишний раз)

3. Это хорошо

4. Для оценки модели все равно же строится несколько траекторий, кажется можно и из них вытащить полезные статистики

File: game/common_methods.py (hunk 3 above)
Comment (krutovsky-danya) [reply game/common_methods.py]: Чтобы строчка не была слишком длинной можно переносить аргументы на новые строки)

Давай для лучшей читаемости все же назовем переменную полно

Comment (Cattharine) [reply game/common_methods.py]: Такой перенос не всегда выглядит приятно

Хорошо

File: environment/__init__.py (hunk 11 above)
Comment (Cattharine) [reply environment/__init__.py:9]: 0. Чудесно)

1-2. Копирование среды - это хорошо, да) Но я решила, что реализовать фиксирование рандома будет быстрее + фиксирование рандома добавляет эффект повторяемости результата (при оценке)

4. Что-то вытащить можно, конечно, но оно будет не сильно точным, думаю

**Review by krutovsky-danya — APPROVED** [review]
